src/net.cpp

```diff
@@ -2813,6 +2831,37 @@ void CNode::AskFor(const CInv& inv)
     mapAskFor.insert(std::make_pair(nRequestTime, inv));
 }
 
+bool CNode::IsAwaitingInternalRequest()
+{
+    return m_block_validation_response.valid();
+}
+
+bool CNode::ProcessInternalRequestResults(NetEventsInterface* peerlogic)
+{
+    bool all_cleared = true;
+
+    if (m_block_validation_response.valid()) {
+        if (m_block_validation_response.wait_for(std::chrono::milliseconds::zero()) == std::future_status::ready) {
+            peerlogic->ProcessBlockValidationResponse(this, m_block_validating, m_block_validating_index, m_block_validation_response.get());
+
+            m_block_validating = nullptr;
+            m_block_validating_index = nullptr;
+            m_block_validation_response = std::future<BlockValidationResponse>();
+        } else {
+            all_cleared = false;
+        }
+    }
+
+    return all_cleared;
+}
+
+void CNode::SetPendingInternalRequest(const std::shared_ptr<const CBlock> block, std::future<BlockValidationResponse>&& pending_response, const CBlockIndex* pindex)
+{
+    m_block_validating = block;
```

Comment:
In commit "Call ProcessNewBlock() asynchronously in a separate thread from p2p layer" (1c9f74170f9c35d7bc5ba3e9e079b70aafae4094)

Would it be possible to assert m_block_validating variables are null/invalid here before overwriting them? It seems like it would help debugging if `SetPendingInternalRequest` were called for a new request before the previous request completed.